uqcsbot/scripts/wolfram.py

```diff
@@ -0,0 +1,192 @@
+            return
+        else:
+            bot.post_message(command.channel, error)
+            return
+
+    # Unpack the response in order to construct the reply message
+    result = wolfram_result['result']  # This is the answer to our question
+    reply_host = wolfram_result['host']  # This is the hostname to ask the next question to
+    conversation_id = wolfram_result['conversationID']  # Used to continue the conversation
+    s_output = wolfram_result.get('s', None)  # s is only sometimes returned but is vital if it is returned
+
+    # TODO: Is there a better option than storing the id in the fallback?
+    # Here we store the conversation ID in the fallback so we can get it back later.
+    # We also store an identifier string to check against later and the reply_host and s_output string
+    attachments = [{
+        'fallback': f'WolframCanReply {reply_host} {s_output} {conversation_id}',
+        'footer': 'Further questions may be asked',
+        'text': result,
+    }]
+
+    bot.post_message(command.channel, "", attachments=attachments)
+
+
+@bot.on('message')
+async def handle_reply(evt: dict):
+    # If the message isn't from a thread or is from a bot ignore it (avoid those infinite loops)
+    if 'thread_ts' not in evt or (evt.get('subtype') == 'bot_message'):
+        return
+
+    channel = evt['channel']
+    thread_ts = evt['thread_ts']  # This refers to time the original message
+    thread_parent = bot.api.conversations.history(channel=channel, limit=1, inclusive=True, latest=thread_ts)
+
+    if not thread_parent['ok']:
+        bot.post_message(channel, 'Sorry, something went wrong with slack', thread_ts=thread_ts)
+        return
+
+    parent_message = thread_parent['messages'][0]  # Limit=1 was used so the first (and only) message is what we want
+    # If the threads parent wasn't by a bot ignore it
+    if parent_message.get('subtype') != 'bot_message':
+        return
+
+    # Finally, we have to check that this is a Wolfram replyable message
+    # It is rare we would reach this point and not pass as who replies to a bot in a thread for another reason?
+    parent_attachments = parent_message['attachments'][0]  # The message is formatted with only one attachment
+    parent_fallback = parent_attachments['fallback']
+    if 'WolframCanReply' not in parent_fallback:
+        return
+
+    # Now we can grab the conversation_id from the message and get the new question (s only sometimes appears)
+    # Recall the format of the fallback "identifier hostname s_output conversationID"
+    _, reply_host, s_output, conversation_id = parent_fallback.split(' ')
+    new_question = evt['text']  # This is the value of the message that triggered this whole response
+    s_output = '' if s_output == 'None' else s_output
+
+    # Slack annoyingly formats the reply_host link so we have to extract what we want:
+    reply_host = reply_host[1:-1].split('|')[0]
+
+    # Now we can ask Wolfram for the next answer:
```

Comment:
I'm seeing a lot of this boilerplate code for making a conversation api call, checking its ok and then extracting the answer-- looks like it can be abstracted into its own function and clear a bit up maybe?